File: python/656.py
"""Project Euler Problem 656: Palindromic sequences.

Define S_alpha(n) = floor(alpha * n) - floor(alpha * (n-1)), and let H_g(alpha) be the sum of the first
g values of n such that S_alpha(1), S_alpha(2), ... S_alpha(n) is a palindrome. Find
sum_beta H_100(sqrt(beta)) for all non-squares beta <= N.

According to https://www.fq.math.ca/Scanned/39-1/komatsu.pdf, the sequence
S_alpha(1) ... S_alpha(n) is a palindrome for n in { 1,2,...q1, q2+q1,2*q2+q1,...q3,
q4+q3,2*q4+q3,...q5, ... }, where q_k are the denominators in the continued
fraction approximants for alpha.
"""

from __future__ import annotations

from math import isqrt


def is_square(n: int) -> bool:
    """Check if n is a perfect square."""
    root = isqrt(n)
    return root * root == n


def cf_sqrt(D: int):
    """Generate continued fraction coefficients for sqrt(D).

    Yields a0, a1, a2, ... (the periodic part repeats).
    """
    a0 = isqrt(D)
    yield a0

    if a0 * a0 == D:
        return

    P = a0
    Q = D - a0 * a0

    while True:
        a = (a0 + P) // Q
        yield a
        P = a * Q - P
        Q = (D - P * P) // Q
        if Q == 0:
            break


def H(beta: int, K: int, M: int) -> int:
    """Compute H_K(sqrt(beta))."""
    qs = [1, 0]
    ns = [0]

    for a in cf_sqrt(beta):
        if len(qs) % 2 == 1:
            for _ in range(a):
                ns.append((ns[-1] + qs[-1]) % M)
                if len(ns) > K:
                    break
        qs.append((qs[-1] * a + qs[-2]) % M)
        if len(ns) > K:
            break

    result = 0
    for i in range(1, min(K + 1, len(ns))):
        result = (result + ns[i]) % M
    return result


def solve() -> int:
    """Solve Problem 656."""
    N = 1000
    K = 100
    M = 10**15

    ans = 0
    for beta in range(1, N + 1):
        if is_square(beta):
            continue
        ans = (ans + H(beta, K, M)) % M

    return ans


def main() -> int:
    """Main entry point."""
    result = solve()
    print(result)
    return result


if __name__ == "__main__":
    main()

File: c/656.c
/*
 * Project Euler 656 - Palindromic sequences
 * H_100(sqrt(beta)) for non-square beta <= 1000.
 * Uses continued fraction of sqrt(D).
 */
#include <stdio.h>
#include <math.h>

typedef long long ll;
typedef unsigned long long ull;

#define M_VAL 1000000000000000LL /* 10^15 */

static int is_square(int n) {
    int r = (int)sqrt((double)n);
    while (r * r > n) r--;
    while ((r + 1) * (r + 1) <= n) r++;
    return r * r == n;
}

static ll H(int beta, int K) {
    int a0 = (int)sqrt((double)beta);
    if (a0 * a0 == beta) return 0;

    /* Generate CF and convergent denominators, collecting palindrome lengths */
    ll ns[110];
    int ns_len = 0;
    ns[ns_len++] = 0;

    ll qs[200]; /* convergent denominators mod M */
    int qs_len = 0;
    qs[qs_len++] = 1;
    qs[qs_len++] = 0;

    int P = a0;
    int Q = beta - a0 * a0;
    int cf_idx = 0; /* 0 = a0 already processed */

    /* Process a0 */
    {
        int a = a0;
        if (qs_len % 2 == 1) { /* odd index: add palindromes */
            for (int rep = 0; rep < a && ns_len <= K; rep++) {
                ns[ns_len] = (ns[ns_len - 1] + qs[qs_len - 1]) % M_VAL;
                ns_len++;
            }
        }
        ll new_q = (qs[qs_len - 1] * a + qs[qs_len - 2]) % M_VAL;
        qs[qs_len++] = new_q;
    }

    /* Continue with periodic part */
    while (ns_len <= K && Q != 0) {
        int a = (a0 + P) / Q;
        if (qs_len % 2 == 1) {
            for (int rep = 0; rep < a && ns_len <= K; rep++) {
                ns[ns_len] = (ns[ns_len - 1] + qs[qs_len - 1]) % M_VAL;
                ns_len++;
            }
        }
        ll new_q = (qs[qs_len - 1] * a + qs[qs_len - 2]) % M_VAL;
        qs[qs_len++] = new_q;

        P = a * Q - P;
        Q = (beta - P * P) / Q;
        if (Q == 0) break;
    }

    ll result = 0;
    int limit = K + 1;
    if (limit > ns_len) limit = ns_len;
    for (int i = 1; i < limit; i++) {
        result = (result + ns[i]) % M_VAL;
    }
    return result;
}

int main(void) {
    int N = 1000;
    int K = 100;

    ll ans = 0;
    for (int beta = 1; beta <= N; beta++) {
        if (is_square(beta)) continue;
        ans = (ans + H(beta, K)) % M_VAL;
    }

    printf("%lld\n", ans);
    return 0;
}
